File: packages/typescript/src/types/messaging.ts
import type { Memory } from "./memory";
import type { Content, UUID } from "./primitives";
import type {
	MessageResult as ProtoMessageResult,
	MessageStreamChunkPayload as ProtoMessageStreamChunkPayload,
	MessageStreamErrorPayload as ProtoMessageStreamErrorPayload,
	TargetInfo as ProtoTargetInfo,
} from "./proto.js";
import type { IAgentRuntime } from "./runtime";

/**
 * Information describing the target of a message.
 */
export interface TargetInfo extends ProtoTargetInfo {
	roomId?: UUID;
	entityId?: UUID;
}

/**
 * Function signature for handlers responsible for sending messages to specific platforms.
 */
export type SendHandlerFunction = (
	runtime: IAgentRuntime,
	target: TargetInfo,
	content: Content,
) => Promise<void>;

export enum SOCKET_MESSAGE_TYPE {
	ROOM_JOINING = 1,
	SEND_MESSAGE = 2,
	MESSAGE = 3,
	ACK = 4,
	THINKING = 5,
	CONTROL = 6,
}

/**
 * WebSocket/SSE event names for message streaming.
 * Used for real-time streaming of agent responses to clients.
 *
 * Event flow:
 * 1. First `messageStreamChunk` indicates stream start
 * 2. Multiple `messageStreamChunk` events with text chunks
 * 3. `messageBroadcast` event with complete message (indicates stream end)
 * 4. `messageStreamError` if an error occurs during streaming
 */
export const MESSAGE_STREAM_EVENT = {
	/** Text chunk during streaming. First chunk indicates stream start. */
	messageStreamChunk: "messageStreamChunk",
	/** Error occurred during streaming */
	messageStreamError: "messageStreamError",
	/** Complete message broadcast (existing event, indicates stream end) */
	messageBroadcast: "messageBroadcast",
} as const;

export type MessageStreamEventType =
	(typeof MESSAGE_STREAM_EVENT)[keyof typeof MESSAGE_STREAM_EVENT];

/**
 * Payload for messageStreamChunk event
 * Uses camelCase for client-facing WebSocket events (JS convention)
 */
export interface MessageStreamChunkPayload
	extends Omit<ProtoMessageStreamChunkPayload, "messageId" | "agentId"> {
	messageId: UUID;
	agentId: UUID;
}

/**
 * Payload for messageStreamError event
 * Uses camelCase for client-facing WebSocket events (JS convention)
 */
export interface MessageStreamErrorPayload
	extends Omit<ProtoMessageStreamErrorPayload, "messageId" | "agentId"> {
	messageId: UUID;
	agentId: UUID;
}

/**
 * Control message actions that can be sent to the frontend
 */
export type ControlMessageAction = "disable_input" | "enable_input";

/**
 * Payload for UI control messages
 */
export interface UIControlPayload {
	/** Action to perform */
	action: ControlMessageAction;
	/** Optional target element identifier */
	target?: string;
	/** Optional reason for the action */
	reason?: string;
	/** Optional duration in milliseconds */
	duration?: number;
}

/**
 * Interface for control messages sent from the backend to the frontend
 * to manage UI state and interaction capabilities
 */
export interface ControlMessage {
	/** Message type identifier */
	type: "control";
	/** Control message payload */
	payload: UIControlPayload;
	/** Room ID to ensure signal is directed to the correct chat window */
	roomId: UUID;
}

/**
 * Handler options for async message processing (User → Agent)
 * Follows the core pattern: HandlerOptions, HandlerCallback, etc.
 */
export interface MessageHandlerOptions {
	/**
	 * Called when the agent generates a response
	 * If provided, method returns immediately (async mode)
	 * If not provided, method waits for response (sync mode)
	 */
	onResponse?: (content: Content) => Promise<void>;

	/**
	 * Called if an error occurs during processing
	 */
	onError?: (error: Error) => Promise<void>;

	/**
	 * Called when processing is complete
	 */
	onComplete?: () => Promise<void>;
}

/**
 * Result of sending a message to an agent (User → Agent)
 * Follows the core pattern: ActionResult, ProviderResult, GenerateTextResult, etc.
 */
export interface MessageResult
	extends Omit<
		ProtoMessageResult,
		"messageId" | "userMessage" | "agentResponses"
	> {
	messageId: UUID;
	userMessage?: Memory;
	agentResponses?: Content[];
}

// ─────────────────────────────────────────────────────────────────────────────
// Database Messaging Types
// ─────────────────────────────────────────────────────────────────────────────

import type { Metadata } from "./primitives";

/**
 * Message Server
 *
 * Represents a messaging platform (Discord, Telegram, etc.) where agents operate.
 * Multiple agents can be associated with a single server.
 */
export interface MessageServer {
	id: UUID;
	name: string;
	sourceType: string;
	sourceId?: string;
	metadata?: Metadata;
	createdAt: Date;
	updatedAt: Date;
}

/**
 * MessagingChannel
 *
 * Represents a conversation space within a message server.
 * Can be a text channel, voice channel, DM, group DM, etc.
 *
 * NOTE: Named "MessagingChannel" (not "Channel") to avoid naming conflicts
 * with ChannelType and other channel-related types.
 */
export interface MessagingChannel {
	id: UUID;
	messageServerId: UUID;
	name: string;
	type: string;
	sourceType?: string;
	sourceId?: string;
	topic?: string;
	metadata?: Metadata;
	createdAt: Date;
	updatedAt: Date;
}

/**
 * MessagingMessage
 *
 * Represents a message sent in a channel (stored in database).
 *
 * NOTE: Named "MessagingMessage" (not "Message" or "DatabaseMessage") to avoid
 * naming conflicts with Memory (which represents agent context) and other
 * message-related types.
 */
export interface MessagingMessage {
	id: UUID;
	channelId: UUID;
	authorId: UUID;
	content: string;
	rawMessage?: Record<string, unknown>;
	sourceType?: string;
	sourceId?: string;
	metadata?: Metadata;
	inReplyToRootMessageId?: UUID;
	createdAt: Date;
	updatedAt: Date;
}

/**
 * Messaging Adapter Interface
 *
 * WHY separate from IDatabaseAdapter: Messaging functionality is specific to
 * certain database backends (SQL adapters) and certain deployment contexts
 * (multi-platform agents). In-memory and local-only adapters don't implement
 * message servers/channels, so these methods cannot be universally provided.
 *
 * WHY this architecture: Following the Interface Segregation Principle - clients
 * should not depend on interfaces they don't use. Client platform plugins
 * (Discord, Telegram) explicitly declare their dependency on IMessagingAdapter,
 * while simple agents can use just IDatabaseAdapter.
 *
 * USAGE: Access via runtime.getMessagingAdapter() or by casting runtime.adapter
 * when you know you're using a SQL backend.
 *
 * @example
 * ```typescript
 * const messagingAdapter = runtime.getMessagingAdapter();
 * if (messagingAdapter) {
 *   const server = await messagingAdapter.createMessageServer({
 *     name: "Discord Server",
 *     sourceType: "discord",
 *     sourceId: "1234567890"
 *   });
 * }
 * ```
 */
export interface IMessagingAdapter {
	// ── Message Server Methods ──────────────────────────────────────────

	/**
	 * Create a new message server
	 *
	 * WHY: When an agent first connects to a platform (Discord, Telegram),
	 * it needs to register that platform as a message server.
	 */
	createMessageServer(data: {
		id?: UUID;
		name: string;
		sourceType: string;
		sourceId?: string;
		metadata?: Metadata;
	}): Promise<MessageServer>;

	/**
	 * Get all message servers
	 */
	getMessageServers(): Promise<MessageServer[]>;

	/**
	 * Get a message server by ID
	 */
	getMessageServerById(serverId: UUID): Promise<MessageServer | null>;

	/**
	 * Get a message server by RLS server ID
	 *
	 * WHY: For Row Level Security (RLS) contexts where server ID is stored
	 * in session variables.
	 */
	getMessageServerByRlsServerId(
		rlsServerId: UUID,
	): Promise<MessageServer | null>;

	/**
	 * Add an agent to a message server
	 *
	 * WHY: A server can have multiple agents (e.g., a Discord server with
	 * multiple bot accounts).
	 */
	addAgentToMessageServer(messageServerId: UUID, agentId: UUID): Promise<void>;

	/**
	 * Get all agent IDs for a message server
	 */
	getAgentsForMessageServer(messageServerId: UUID): Promise<UUID[]>;

	/**
	 * Remove an agent from a message server
	 */
	removeAgentFromMessageServer(
		messageServerId: UUID,
		agentId: UUID,
	): Promise<void>;

	// ── Channel Methods ─────────────────────────────────────────────────

	/**
	 * Create a new channel
	 *
	 * WHY: When the agent joins/creates a channel on a platform, it needs to
	 * store the channel metadata for future message routing.
	 *
	 * @param data Channel properties
	 * @param participantIds Optional initial participant list
	 */
	createChannel(
		data: {
			id?: UUID;
			messageServerId: UUID;
			name: string;
			type: string;
			sourceType?: string;
			sourceId?: string;
			topic?: string;
			metadata?: Metadata;
		},
		participantIds?: UUID[],
	): Promise<MessagingChannel>;

	/**
	 * Get all channels for a message server
	 */
	getChannelsForMessageServer(
		messageServerId: UUID,
	): Promise<MessagingChannel[]>;

	/**
	 * Get channel details by ID
	 */
	getChannelDetails(channelId: UUID): Promise<MessagingChannel | null>;

	/**
	 * Update channel properties
	 */
	updateChannel(
		channelId: UUID,
		updates: {
			name?: string;
			participantCentralUserIds?: UUID[];
			metadata?: Metadata;
		},
	): Promise<MessagingChannel>;

	/**
	 * Delete a channel
	 */
	deleteChannel(channelId: UUID): Promise<void>;

	/**
	 * Add participants to a channel
	 *
	 * WHY: When users join a channel, they need to be tracked as participants
	 * for permission checks and message delivery.
	 */
	addChannelParticipants(channelId: UUID, entityIds: UUID[]): Promise<void>;

	/**
	 * Get all participant IDs for a channel
	 */
	getChannelParticipants(channelId: UUID): Promise<UUID[]>;

	/**
	 * Check if an entity is a channel participant
	 */
	isChannelParticipant(channelId: UUID, entityId: UUID): Promise<boolean>;

	// ── Message Methods ─────────────────────────────────────────────────

	/**
	 * Create a new message
	 *
	 * WHY: When a message is received from a platform or sent by the agent,
	 * it's stored for conversation history, context, and retrieval.
	 */
	createMessage(data: {
		channelId: UUID;
		authorId: UUID;
		content: string;
		rawMessage?: Record<string, unknown>;
		sourceType?: string;
		sourceId?: string;
		metadata?: Metadata;
		inReplyToRootMessageId?: UUID;
		messageId?: UUID;
	}): Promise<MessagingMessage>;

	/**
	 * Get a message by ID
	 */
	getMessageById(id: UUID): Promise<MessagingMessage | null>;

	/**
	 * Update a message
	 *
	 * WHY: Messages can be edited after being sent (e.g., Discord edit events).
	 */
	updateMessage(
		id: UUID,
		patch: {
			content?: string;
			rawMessage?: Record<string, unknown>;
			sourceType?: string;
			sourceId?: string;
			metadata?: Metadata;
			inReplyToRootMessageId?: UUID;
		},
	): Promise<MessagingMessage | null>;

	/**
	 * Get messages for a channel with pagination
	 *
	 * WHY: Loading conversation history for context or display.
	 *
	 * @param channelId The channel to fetch messages from
	 * @param limit Max messages to return (default 50)
	 * @param beforeTimestamp Get messages before this timestamp (for pagination)
	 */
	getMessagesForChannel(
		channelId: UUID,
		limit?: number,
		beforeTimestamp?: Date,
	): Promise<MessagingMessage[]>;

	/**
	 * Delete a message
	 */
	deleteMessage(messageId: UUID): Promise<void>;

	/**
	 * Find or create a DM channel between two users
	 *
	 * WHY: Direct message channels are created on-demand when two users
	 * start a conversation. Ensures we don't create duplicate DM channels.
	 */
	findOrCreateDmChannel(
		user1Id: UUID,
		user2Id: UUID,
		messageServerId: UUID,
	): Promise<MessagingChannel>;
}

File: packages/rust/src/types/messaging.rs
//! Messaging types for elizaOS
//!
//! Contains messaging-related types for agent communication.

use serde::{Deserialize, Serialize};

use super::primitives::{Content, UUID};

/// Target information for sending messages
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TargetInfo {
    /// Target type
    #[serde(rename = "type")]
    pub target_type: TargetType,
    /// Room ID (for room targets)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub room_id: Option<UUID>,
    /// Entity ID (for entity targets)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity_id: Option<UUID>,
    /// Channel ID (for external platforms)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub channel_id: Option<String>,
    /// Platform source (e.g., "discord", "telegram")
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
}

/// Target type for message sending
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TargetType {
    /// Send to a room
    Room,
    /// Send to an entity
    Entity,
    /// Send to a channel
    Channel,
}

/// Control message for runtime control
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ControlMessage {
    /// Message type
    #[serde(rename = "type")]
    pub message_type: ControlMessageType,
    /// Message data
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

/// Control message types
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ControlMessageType {
    /// Pause processing
    Pause,
    /// Resume processing
    Resume,
    /// Stop the agent
    Stop,
    /// Restart the agent
    Restart,
    /// Update configuration
    UpdateConfig,
}

/// Message queue item
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageQueueItem {
    /// Message ID
    pub id: UUID,
    /// Room ID
    pub room_id: UUID,
    /// Message content
    pub content: Content,
    /// Priority (higher = more urgent)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub priority: Option<i32>,
    /// Creation timestamp
    pub created_at: i64,
    /// Processing status
    pub status: MessageQueueStatus,
}

/// Message queue status
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MessageQueueStatus {
    /// Waiting to be processed
    Pending,
    /// Currently being processed
    Processing,
    /// Successfully processed
    Completed,
    /// Failed to process
    Failed,
}

/// Send handler function type (represented as a name for serialization)
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SendHandlerInfo {
    /// Handler source/platform
    pub source: String,
    /// Whether the handler is active
    pub active: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_target_info_serialization() {
        let target = TargetInfo {
            target_type: TargetType::Room,
            room_id: Some(UUID::new_v4()),
            entity_id: None,
            channel_id: None,
            source: Some("discord".to_string()),
        };

        let json = serde_json::to_string(&target).unwrap();
        assert!(json.contains("\"type\":\"room\""));
        assert!(json.contains("\"source\":\"discord\""));
    }

    #[test]
    fn test_control_message_serialization() {
        let msg = ControlMessage {
            message_type: ControlMessageType::Pause,
            data: None,
        };

        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"PAUSE\""));
    }
}
